cdap-data-fabric/src/main/java/co/cask/cdap/data2/registry/UsageRegistry.java

```diff
@@ -57,7 +58,62 @@ public UsageDatasetIterable get() {
   }
 
   /**
+   * Registers usage of a stream by multiple ids.
+   *
+   * @param users the users of the stream
+   * @param streamId the stream
+   */
+  public void registerAll(List<Id> users, Id.Stream streamId) {
+    for (Id user : users) {
```

Comment:
Shouldn't be done in a loop like this. Each call to register open a new transaction and write to store. It should be just one transaction, one batch write.